### PR TITLE
Allow multi-line names for spectre test cases

### PR DESCRIPTION
**Pull request by knelli2**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: cmake/SpectreParseTests.py [hunk 1]
```diff
@@ -101,8 +101,18 @@ def parse_source_file(file_name):
         # "Unit.My.Test", "[Unit][My]"
         # group(1) == Unit.My.Test
         # group(2) == [Unit][My]
-        parsed_name = re.search(r'"(.*)",[\s]*"(.*)"', test_name)
-        test_name = parsed_name.group(1)
+        #
+        # Also handle cases where the name is very long and goes onto multiple
+        # lines (this requires the name to have parentheses because of how C++
+        # macros work),
```
Comment (wthrowe) [inline cmake/SpectreParseTests.py]: I don't understand this comment.  That's not something I would expect the preprocessor to care about.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Can you give an example of how to use this?  I can't get it to work.

File: cmake/SpectreParseTests.py (hunk 1 above)
Comment (knelli2) [reply cmake/SpectreParseTests.py]: This is the use case I found that didn't work (on a branch, not develop):

```cpp
SPECTRE_TEST_CASE(("Unit.ParallelAlgorithms.Interpolation.Runtime.Callbacks."
                   "ObserveTimeSeriesOnSurface"),
                  "[Unit]") {
  test_option_parsing();
}
```

When I didn't have the `()` surrounding the test name, CMake complained. Also in other instances of macros, I found that I sometimes had to wrap a multi-line argument in `()`

Comment (wthrowe) [reply cmake/SpectreParseTests.py]: If CMake is complaining that's not a problem with macros.  CMake isn't running the preprocessor.

Have you checked that your test actually runs?  If I try something like what you have the test is silently ignored.

Comment (knelli2) [reply cmake/SpectreParseTests.py]: Yup my test runs

Comment (wthrowe) [reply cmake/SpectreParseTests.py]: Did you re-run cmake?  If I copy-paste your example into a file it's silently ignored.

Comment (knelli2) [reply cmake/SpectreParseTests.py]: Ok now it works with the test case

```cpp
SPECTRE_TEST_CASE("Unit.ParallelAlgorithms.Interpolation.Runtime.Callbacks."
                  "ObserveTimeSeriesOnSurface",
                  "[Unit]") {
  test_option_parsing();
}
```

I wasn't actually running the test through `ctest`. I was just running the executable directly and the test did run which is why I thought it had worked. Now I checked with `ctest`

Comment (nilsdeppe) [reply cmake/SpectreParseTests.py]: I think the regex I wrote in CMake is probably the part that's failing (or maybe it's regex in python, I forget what parses these nowadays). Just an FYI

Comment (knelli2) [reply cmake/SpectreParseTests.py]: Turned out to be both :)

File: cmake/SpectreAddCatchTests.cmake [hunk 2]
```diff
@@ -143,6 +143,8 @@ function(spectre_parse_file SOURCE_FILE TEST_TARGET)
     # Strip wrapping quotation marks of each element of the list
     # TEST_STRINGS
     string(REGEX REPLACE "^\"(.*)\"$" "\\1" TEST_STRINGS "${TEST_STRINGS}")
+    string(REPLACE ".\";\"" "." TEST_STRINGS "${TEST_STRINGS}")
+    string(REPLACE "\";\"." "." TEST_STRINGS "${TEST_STRINGS}")
```
Comment (wthrowe) [inline cmake/SpectreAddCatchTests.cmake:150]: Update the comment on this block of replacements or put the ones to join the list entries in a separate block.

File: cmake/SpectreParseTests.py (hunk 1 above)
Comment (wthrowe) [reply cmake/SpectreParseTests.py]: Seems to work without parentheses now, but it is silently ignored with parentheses.  This is probably fine, but remove anything related to handling parentheses.

File: cmake/SpectreParseTests.py [hunk 3]
```diff
@@ -101,8 +101,18 @@ def parse_source_file(file_name):
...
+        # ("Unit.My.Test.Name."
+        #     "Is.Very.Long"),
+        # "[Unit][My]"
+        # group(1) == Unit.My.Test.Name.Is.Very.Long
```
Comment (wthrowe) [inline cmake/SpectreParseTests.py]: `group(1)` includes some quotes and whitespace.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by knelli2** [comment]
Squashed all the changes because they were short

**Review by wthrowe — APPROVED** [review]

**Pull request merged**